### PR TITLE
feat: add COSI test suite using Chainsaw

### PR DESCRIPTION
**Pull request by shanduur**
This pull request introduces a new COSI (Container Object Storage Interface) test suite using Chainsaw, along with several enhancements and preparations for dynamic test generation.

1. **COSI Test Suite Implementation**:
   - Added an end-to-end (E2E) test suite for COSI, focusing on various authentication methods for S3 and Azure.

2. **Test Generator**:
   - Implemented a test generator to dynamically create tests, reducing manual effort and ensuring consistency across the test suite.

3. **KinD Cluster Configuration**:
   - Configured [`ctlptl`](https://github.com/tilt-dev/ctlptl) for setting up a KinD (Kubernetes in Docker) cluster to facilitate testing in a controlled environment. 
   **TODO:** this needs future updates, e.g. Makefile targets, to ensure that tests can be run locally, including installation of sample driver, CRDs and controller.

The dynamic test generator is now in place, which will allow for easier updates and extensions to the test suite in future iterations. This PR is foundational for future enhancements and ensures that the COSI functionality is thoroughly tested across different authentication mechanisms.

### CODE REVIEW

File: test/generator/config-sample.yaml
[the remark this answers is not in the document]
Comment (shanduur) [reply test/generator/config-sample.yaml:1]: I have no idea if my change will work at all, I still need to test that.

**Comment by BlaineEXE** [comment]
In this other PR, I'm adding the 3 remaining e2e/lint targets that are suggested by Kubebuilder's template: https://github.com/kubernetes-sigs/container-object-storage-interface-api/pull/96/files

I think it would be nice to see how this PR would integrate with make in the `test-e2e` target. Presumably, the tooling would have to install ctlptl and chainsaw, set up kind, and then run the tests, which seems fairly straighforward.

If you think it would be better to do the tooling in another PR, maybe a `test/e2e/readme.md` could give a simple bullet list instruction about how we can run these ourselves manually?

**Comment by shanduur** [comment]
I fully tested that on Linode COSI Driver using the following values:

```yaml
driverName: objectstorage.cosi.linode.com
deletionPolicy: "Delete" # "Retain"
bucketClassParams:
  cosi.linode.com/v1/region: us-east
  cosi.linode.com/v1/acl: private
  cosi.linode.com/v1/cors: disabled
bucketAccessClassParams:
  cosi.linode.com/v1/permissions: read_write
authenticationType: "Key" # "IAM"
bucketClaimProtocols: ["S3"]
bucketAccessProtocol: "S3"
```

Test suite run in ~22s, including full `Bucket`/`BucketAccess`/`BucketClaim` provisioning and secret schema validation.

**Comment by shanduur** [comment]
I included a simple readme, however, I am not adding any `make` targets yet. I think it will be better to do those in subsequent PRs.

**Review by BlaineEXE — APPROVED** [review]

**Comment by BlaineEXE** [comment]
Looks like a great starting point to me. Thanks @shanduur !

**Pull request merged**